examples/run_summarization_finetuning.py

```diff
@@ -0,0 +1,625 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Finetuning seq2seq models for sequence generation."""
+
+import argparse
+from collections import deque
+import logging
+import os
+import pickle
+import random
+import sys
+
+import numpy as np
+from tqdm import tqdm, trange
+import torch
+from torch.optim import Adam
+from torch.utils.data import Dataset, DataLoader, RandomSampler, SequentialSampler
+
+from transformers import AutoTokenizer, PreTrainedSeq2seq, Model2Model
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+
+
+def set_seed(args):
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+
+
+# ------------
+# Load dataset
+# ------------
+
+
+class TextDataset(Dataset):
+    """ Abstracts the dataset used to train seq2seq models.
+
+    CNN/Daily News:
+
+    The CNN/Daily News raw datasets are downloaded from [1]. The stories are
+    stored in different files; the summary appears at the end of the story as
+    sentences that are prefixed by the special `@highlight` line. To process
+    the data, untar both datasets in the same folder, and pass the path to this
+    folder as the "data_dir argument. The formatting code was inspired by [2].
+
+    [1] https://cs.nyu.edu/~kcho/
+    [2] https://github.com/abisee/cnn-dailymail/
+    """
+
+    def __init__(self, tokenizer, prefix="train", data_dir="", block_size=512):
+        assert os.path.isdir(data_dir)
+
+        # Load features that have already been computed if present
+        cached_features_file = os.path.join(
+            data_dir, "cached_lm_{}_{}".format(block_size, prefix)
+        )
+        if os.path.exists(cached_features_file):
+            logger.info("Loading features from cached file %s", cached_features_file)
+            with open(cached_features_file, "rb") as source:
+                examples = pickle.load(source)
+                self.source_examples = examples["source"]
+                self.target_examples = examples["target"]
+                return
+
+        logger.info("Creating features from dataset at %s", data_dir)
+        self.source_examples = []
+        self.target_examples = []
+        datasets = ["cnn", "dailymail"]
+        for dataset in datasets:
+            path_to_stories = os.path.join(data_dir, dataset, "stories")
+            assert os.path.isdir(path_to_stories)
```

Comment:
the goal of the assertion is not clear since `listdir` will raise an exception on an invalid path.